src/kotlin/kage/crypto/chacha20/ChaCha20Poly1305OutputStream.kt

```diff
@@ -0,0 +1,75 @@
+package kage.crypto.chacha20
+
+import kage.crypto.chacha20.ChaCha20Poly1305.CHACHA_20_POLY_1305_NONCE_LENGTH
+import java.io.OutputStream
+
+// TODO: This is not thread safe
+internal class ChaCha20Poly1305OutputStream(private val key: ByteArray, private val dst: OutputStream) : OutputStream() {
+
+  internal companion object {
+    const val CHUNK_SIZE = 64 * 1024
+  }
+
+  private val nonce = ByteArray(CHACHA_20_POLY_1305_NONCE_LENGTH)
+
+  private val buf = ByteArray(CHUNK_SIZE)
+  private var bufSize = 0
+
+  override fun write(i: Int) {
+    val b = i.toByte()
+
+    buf[bufSize] = b
+    bufSize++
+
+    if(bufSize == CHUNK_SIZE) {
+      flushChunk(last = false)
+      bufSize = 0
+    }
+  }
+
+  override fun close() {
+    flushChunk(last = true)
+    dst.close() // TODO: Should we close the underlying stream?
+  }
+
+  /*
+  This will encrypt an empty block when the last buffer is empty. That is, when the playload size is a multiple of
+  `chunkSize`, the last encrypted block will contain an empty payload, encrypted using a nonce with the `last` flag set.
+  to 1.
+
+  The spec explicitly says this must not happen:
+
+  > The final chunk MAY be shorter than 64 KiB but MUST NOT be empty unless the whole payload is empty.
+
+  However, the go implementation does the same as this code and getting around this would be tricky as we'd have to go
+  back after writing and encrypt the last chunk again with a different nonce, with the `last` flag set to 1.
+
+  Maybe the spec means the last **encrypted** chunk must not be empty?
+   */
```

Comment:
the relevant line in the spec: https://github.com/C2SP/C2SP/blob/main/age.md?plain=1#L146